sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -30,7 +30,7 @@ import org.apache.spark.sql.internal.SQLConf
 case class CoalesceShufflePartitions(session: SparkSession) extends CustomShuffleReaderRule {
 
   override val supportedShuffleOrigins: Seq[ShuffleOrigin] =
-    Seq(ENSURE_REQUIREMENTS, REPARTITION_BY_COL, REPARTITION_BY_NONE)
+    Seq(ENSURE_REQUIREMENTS, REPARTITION_BY_NONE, REPARTITION_BY_COL, REBALANCE_PARTITIONS)
```

Comment:
Do we need to update `OptimizeLocalShuffleReader`?

Reply:
We can't do this. Otherwise the columns cann't be partitioned and dynamic partition writing will produce many small files.

Reply:
Can we update `OptimizeLocalShuffleReader` a bit more to not optimize `HashPartitioning`? Ideally `REBALANCE_PARTITIONS` without parameters should be able to apply local shuffle reader.

Reply:
split `REBALANCE_PARTITIONS` shuffle origin into two parts which is similr with `REPARTITION` ?
* REBALANCE_PARTITIONS_BY_COL
* REBALANCE_PARTITIONS_BY_NONE

Reply:
ah, SGTM! We can probably replace `REPARTITION_BY_NONE` with `REBALANCE_PARTITIONS_BY_NONE`

Reply:
yea, this also resolve my previous comment https://github.com/apache/spark/pull/32932#discussion_r652649568
FYI @viirya